cardano-crypto-class/src/Cardano/Crypto/Hash/Blake2b.hs

```diff
@@ -14,24 +17,128 @@ where
 import Control.Monad (unless)
 import Cardano.Crypto.Libsodium.C (c_crypto_generichash_blake2b)
 
-import Cardano.Crypto.Hash.Class (HashAlgorithm, SizeHash, hashAlgorithmName, digest)
+import Cardano.Crypto.Hash.Class (HashAlgorithm (..), SizeHash, hashAlgorithmName, digest)
 import Foreign.Ptr (castPtr, nullPtr)
 import Foreign.C.Error (errnoToIOError, getErrno)
 import GHC.IO.Exception (ioException)
+import qualified Data.ByteString.Short.Internal as SBSI
+import qualified Data.ByteString.Short as SBS
+import Data.ByteString.Short (ShortByteString)
+import Data.Word (Word64, Word32)
 
 import qualified Data.ByteString as B
 import qualified Data.ByteString.Internal as BI
+import           Control.DeepSeq (NFData)
+import           NoThunks.Class (NoThunks)
+import GHC.Generics (Generic)
+import qualified Data.Primitive.ByteArray as BA
+import qualified Control.Monad.ST as ST
 
 data Blake2b_224
 data Blake2b_256
 
+data PackedBytes32 =
+  PackedBytes32
+    {-# UNPACK #-} !Word64
+    {-# UNPACK #-} !Word64
+    {-# UNPACK #-} !Word64
+    {-# UNPACK #-} !Word64
```

Comment:
You can save here an extra 4 bytes if you use: [`Word32#`](https://downloads.haskell.org/~ghc/8.10.4/docs/html/libraries/base-4.14.1.0/GHC-Exts.html#t:Word32-35-) for `Blake2b_224`

You'd have to use indexing and writing from `GHC.Exts` though, since those aren't default types until ghc-9.2

* [`indexWord8ArrayAsWord32#`](https://downloads.haskell.org/~ghc/8.10.4/docs/html/libraries/base-4.14.1.0/GHC-Exts.html#v:indexWord8ArrayAsWord32-35-)
* [`writeWord8ArrayAsWord32#`](https://downloads.haskell.org/~ghc/8.10.4/docs/html/libraries/base-4.14.1.0/GHC-Exts.html#v:writeWord8ArrayAsWord32-35-)